google_guest_agent/events/events.go

```diff
@@ -84,224 +150,336 @@ type WatcherEventType struct {
-	ep.mutex.Lock()
-	defer ep.mutex.Unlock()
-	delete(ep.watchersMap, evType)
-	return len(ep.watchersMap)
+// AddWatcher adds/enables a new watcher. The watcher will be fired up right away if the
+// event manager is already running, otherwise it's scheduled to run when Run() is called.
+func (mngr *Manager) AddWatcher(ctx context.Context, watcher Watcher) error {
+	mngr.watchersMutex.Lock()
+	defer mngr.watchersMutex.Unlock()
+	id := watcher.ID()
+	if _, found := mngr.watchersMap[id]; found {
+		return fmt.Errorf("Watcher(%s) was previously added", id)
+	}
+
+	// Add the watchers and its events to internal mappings.
+	evTypes := make(map[string]*WatcherEventType)
+	mngr.watchersMap[id] = true
+
+	for _, curr := range watcher.Events() {
+		evType := &WatcherEventType{
+			watcher: watcher,
+			evType:  curr,
+			removed: make(chan bool),
+		}
+
+		evTypes[curr] = evType
+		mngr.watcherEvents = append(mngr.watcherEvents, evType)
+	}
+
+	// If we are not running don't bother "running" the watcher, Run() will do it later.
+	if !mngr.running {
+		return nil
+	}
+
+	// If we are already running the "run/launch" the watcher.
+	for _, curr := range watcher.Events() {
+		logger.Debugf("Adding watcher for event: %s", curr)
+		mngr.queue.add(curr)
+		go func(watcher Watcher, evType string, removed chan bool) {
+			mngr.runWatcher(ctx, watcher, evType, removed)
+		}(watcher, curr, evTypes[curr].removed)
+	}
+
+	return nil
+}
+
+func (mngr *Manager) runWatcher(ctx context.Context, watcher Watcher, evType string, removed chan bool) {
+	nCtx, cancel := context.WithCancel(ctx)
+	abort := false
+	id := watcher.ID()
+
+	go func() {
+		abort = <-removed
+		logger.Debugf("Got a request to abort watcher(%s) for event: %s", id, evType)
+		cancel()
+	}()
+
+	for renew := true; renew; {
+		var evData interface{}
+		var err error
+
+		renew, evData, err = watcher.Run(nCtx, evType)
+
+		logger.Debugf("Watcher(%s) returned event: %q, should renew?: %t", id, evType, renew)
+
+		if abort || mngr.queue.leaving {
+			logger.Debugf("Watcher(%s), either are aborting(%t) or leaving(%t), breaking renew cycle",
+				id, abort, mngr.queue.leaving)
+			break
+		}
+
+		mngr.queue.dataBus <- eventBusData{
+			evType: evType,
+			data: &EventData{
+				Data:  evData,
+				Error: err,
+			},
+		}
+	}
+
+	logger.Debugf("watcher finishing: %s", evType)
+	if !abort {
+		removed <- true
+	}
+
+	mngr.queue.watcherDone <- evType
+}
+
+func (mngr *Manager) getSubscribers(evType string) []*eventSubscriber {
+	var res []*eventSubscriber
+
+	mngr.subscribersMutex.Lock()
+	defer mngr.subscribersMutex.Unlock()
+
+	tmpSlice, found := mngr.subscribers[evType]
+	if !found {
+		return nil
+	}
+
+	for _, curr := range tmpSlice {
+		res = append(res, curr)
+	}
+
+	return res
 }
 
 // Run runs the event manager, it will block until all watchers have given up/failed.
-func (mngr *Manager) Run(ctx context.Context) {
+func (mngr *Manager) Run(ctx context.Context) error {
```

Comment:
on return shouldn't this be closing all channels (like new one's introduced in `queue` structure) its listening on?